### PR TITLE
365 tensorflow upgrade nn hack

### PR DESCRIPTION
**Pull request by danieltudosiu**
## Status
READY

## Description
NiftyNet was tested if it works with Tensorflow 1.13.1/1.14.0rc0 and CUDA 10.0.130 with cudnn 7.5.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)
- [x] My code follows the code style of this project.
- [x] All new and existing tests passed.

## Todos
- [x] Tests
- [x] Documentation

## Impacted Areas in Application
List general components of the application that this PR will affect:
niftynet
    |
    -layer
    |   |
    |   -crf.py
    |
    -contrib
    |   |
    |   -segmentation_selective_sampler
    |       |
    |       -test_sampler_selective.py
    -io
        |
        -misc_io.py
tests
    |
    -most tests (too lazy to put 25+ files in the nice format I used)

### CODE REVIEW

File: tests/niftynet_testcase.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import tensorflow as tf
+
+try:
+    delattr(tf.test.TestCase,'test_session')
+except AttributeError:
+    pass
+
+
+class NiftyNetTestCase(tf.test.TestCase):
+    pass
```
Comment (ericspod) [inline tests/niftynet_testcase.py:17]: Newline here.

File: tests/niftynet_testcase.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import tensorflow as tf
+
+try:
+    delattr(tf.test.TestCase,'test_session')
```
Comment (ericspod) [inline tests/niftynet_testcase.py:11]: Add a comment here stating why we've had to do this.

File: niftynet/io/misc_io.py [hunk 3]
```diff
@@ -779,10 +783,14 @@ def close_logger():
 
     :return:
     """
-    # pylint: disable=no-name-in-module
-    from tensorflow.python.platform.tf_logging import _get_logger
+    try:
```
Comment (ericspod) [inline niftynet/io/misc_io.py:807]: I would comment here that this is needed for compatibility reasons between TF versions

File: requirements-gpu.txt [hunk 4]
```diff
@@ -1,9 +1,9 @@
-six>=1.10
-nibabel>=2.1.0
-numpy>=1.13.3, <= 1.14.5
-scipy>=0.18
+six>=1.12.0
+nibabel>=2.4.1
+numpy>=1.16.4
```
Comment (ericspod) [inline requirements-gpu.txt]: I would have numpy>=1.13.3

File: requirements-gpu.txt [hunk 5]
```diff
@@ -1,9 +1,9 @@
...
+scipy>=1.2.1
 configparser
-tensorflow-gpu==1.12.2
+tensorflow-gpu==1.14.0rc0
```
Comment (ericspod) [inline requirements-gpu.txt:6]: I would have >=1.12.2

**Review by ericspod — CHANGES_REQUESTED** [review]
Other than adding a few more comments I would say this is good to go.

**Review by ericspod — CHANGES_REQUESTED** [review]
In the added comments, "brake" -> "break" but otherwise good. There are conflicts with other tests that were modified today before the PR was made. The tests are also failing because CUDA 10  isn't getting loaded on the host VM. We need to set the requirements for now to be TF 1.12 as they were before, we know that the test will work on later versions and will have to return to upgrading the test host.

**Comment by danieltudosiu** [comment]
@ericspod what are we doing with this pull request?

**Comment by danieltudosiu** [comment]
So what should we do based on the Tensorflow Issue [Unittest and test_session interaction #29439](https://github.com/tensorflow/tensorflow/issues/29439) response? Change the tests naming scheme so we do not use the non-OOP fix?

**Comment by ericspod** [comment]
It wasn't our names it was theirs that was the issue. I say we keep our
delattr solution for now.

Eric

**Review by ericspod — APPROVED** [review]

**Pull request merged**